### ISSUE TITLE
Refactor HTML/CSS patterns in UI Library

### ISSUE THREAD

**Issue by NateWr** [pkp/pkp-lib#7499]
**Describe the problem you would like to solve**
The [UI Library](https://docs.pkp.sfu.ca/dev/ui-library/dev) uses the [BEM](http://getbem.com/introduction/) methodology for writing HTML/CSS based around components. In large projects, this reduces the frequency with which a change in one CSS rule impacts other components in unexpected ways.

However, it also leads to a lot of duplicated effort when bringing components together into a full page layout. How components are positioned, the space between elements like buttons and titles, and the typography are all defined with CSS rules written specific to each component or page. This leads to two problems:

1. The typography, positioning and spacing of components can be inconsistent on different pages.
2. Developers who are not great at design must try to determine the correct positioning or spacing of components themselves.

**Describe the solution you'd like**
Adapt the HTML/CSS patterns to use utility classes similar to [Tailwind](https://tailwindcss.com/). The goal is not to replace BEM with the utility-first approach of Tailwind. Both approaches should be combined in the following way:

1. Use BEM to apply styles within a component. For example, the [Filter](https://docs.pkp.sfu.ca/dev/ui-library/dev/#/component/Filter) component would use BEM classes to style the currently selected filter and show the X button.
2. Use utility classes to apply typography modifiers (size/weight) and to position components together on a page (margin/padding/flex positioning/grids).

In addition to these utility classes, the UI Library should adopt "Elements", which are HTML/CSS patterns that are not specifically Vue.js components. These could be used, for example, to apply common styles to `<table>` or `<ul>` components that shouldn't be bound to a Vue.js component.
 
**Who is asking for this feature?**
As the maintainer of the UI Library, I feel this is important to ensure the library can be more easily learned and used by all members of the team.

**Comment by jonasraoni** [comment]
One of the interesting things about such "utility classes" like Tailwind is the ability to copy and paste components =]
But sometimes it leads to a bloated salad of classes, so I also support using both together, while trying to give preference to the utility classes.

**Comment by NateWr** [comment]
> sometimes it leads to a bloated salad of classes, so I also support using both together

Yes! My plan is to limit their use to things like margins, padding, font weight, and layout. Basically, where styles should be _consistent_ across components (spacing, font size and weights, etc), we can introduce a utility class.

**Comment by jardakotesovec** [comment]
Hi Everyone,

I have been thinking about this for a while. And I agree that utility classes are great for capturing some design patterns, including spacing, typography and colors.

I have used tailwindcss in some previous projects (not my choice - and still really liked it!). It has become very popular framework which I believe has strong future. Usually very well received especially with developers who are not that familiar with css. Its more approachable than pure css. It would be probably contra-productive trying to do something own that would be similar.

I am worried that if we would aim to combining BEM and utility classes - that it basically requires developers to be good at both and also know where to use what. Thats why I am inclined at this point to just go full tailwinds.

I created example tailwind configuration based mostly on @Devika008 design pattern that was introduced in new submission listing (https://docs.google.com/spreadsheets/d/1S-gJUSw_hZyiANiKitxP_XOVgQPAgIa-PBSJNZFWPjE/edit#gid=0) - https://github.com/pkp/ojs/pull/4064/files . Its just initial version, it would get refined while I would be using that to build new submission listing. I think seeing the config helps to see how tailwind can be useful in expressing design patterns. 

Obviously biggest downside to go this route is that we would need to update lots of existing styling - on the other hand it might be good opportunity to do the UI refinement which would be aligned with current @Devika008  designs. 

Any technology transition I think should have time frame and I think its realistic to aim that with releasing 3.6 our admin interface could be pure tailwindcss.

I am intentionally sharing this in very early stages so there is still opportunity for push back and feedback. So please tell me what you think.

Thank you!

**Comment by Vitaliy-1** [comment]
Personally I like Bootstrap approach for web development/naming conventions and don't like BEM at all. I'd agree with Jonas and Nate to avoid oversaturating the code with repeated names of HTML classes. We could hardcode the repeated patterns in CSS or use other approaches. I suspect tailwind also offers something?

**Comment by jonasraoni** [comment]
Two years later... After having a small experience maintaining a design system, I'd go with a ready set of components and use whatever it's running under it (Bootstrap, Tailwind, ...). Also, I would definitely avoid customizing/extending its internals, so that I can update things easily and benefit from its skins/upgrades.

Off the top of my head, I just remember about accessible React components (e.g. https://react-spectrum.adobe.com/react-aria), but there must be something interesting for Vue.

But that's just my opinion :)

**Comment by jardakotesovec** [comment]
@jonasraoni Yeah, react-aria is my go to framework when I want to see some good accessibility implementations. They have probably most thorough accessibility UI layers you can find. And they are styling agnostic - so whether you use tailwind or pure css - up to you.

Unfortunately there is nothing even close on vue.js side. https://headlessui.com is pretty good, so I used that for some things when moving to vue3.  But its not really evolving.  Its from tailwindcss team and you can see that they are actually leveraging react-aria on new projects.. and vue.js is falling behind at this point.
But its important to realise that even in React.js ecosystem these are pretty new and evolving solutions. It turned out that react-aria is not very popular because its hard to use - so they came up with react aria components which has lot more friendly API. But thats also still quite new and evolving.

We are here for many years.. and if anything we want to adopt solutions that are converging and won't be abandoned in 18months. And I can see that the frontend world is heavily moving towards headless components, but its still evolving quite quickly. My hope is that react.js explores different approaches and vue.js copies the best one (there is some attempt https://github.com/radix-vue/radix-vue). Similar story was with store architecture (all the redux/flux stuff in react vs one go to solution pinia in vue.js) 

Our components have lots of business specific functionality. We can't just swap our up-library for something else. But what make sense to don't spend time on maintaining JS&accessibility functionality for patterns that has been implemented many times (autocomplete, modal, tooltips, menus, ...) and for that case these headless libraries are way to go. And I am closely watching space for that to offload when possible.

**Comment by jardakotesovec** [comment]
@Vitaliy-1 Have not used bootstrap for a while, but they seems to be moving to very similar direction as tailwind with directly putting utility classes to class names - https://getbootstrap.com/docs/5.3/utilities/display/#examples .

Repeating patterns should not really happen if we break things down reasonably to the vue.js components which we should do anyway.

**Comment by Vitaliy-1** [comment]
> @Vitaliy-1 Have not used bootstrap for a while, but they seems to be moving to very similar direction as tailwind with directly putting utility classes to class attribute - https://getbootstrap.com/docs/5.3/utilities/display/#examples .

Yeah, they have been using this. As far as I remember it's used more in the way as a "helper" alongside with their components library. Basically, how I would preferred.

> Repeating patterns should not really happen if we break things down reasonably to the vue.js components which we should do anyway.

I saw several examples with Livewire/Blade components and can agree that this might work with Vue.

**Comment by jonasraoni** [comment]
> (...) Our components have lots of business specific functionality. We can't just swap our ui-library for something else. But what make sense to don't spend time on maintaining JS&accessibility functionality for patterns that has been implemented many times (autocomplete, modal, tooltips, menus, ...) and for that case these headless libraries are way to go. And I am closely watching space for that to offload when possible.

I totally agree with this vision.

Then about choosing a tool/library (e.g. Tailwind), there are some things to analyze, but it's always a guess at what will survive the test of time, so I don't have anything against it. Just wanted to share my view which fits what you've wrote 😁

**Comment by jardakotesovec** [comment]
For 3.5 we adopted tailwindcss for everything new in ui-library we create or refactor, which made straightforward to express our design system just with the config and so far it feels like a good fit for editorial backend. Any repeating patterns are captured by new vue.js components. More details is captured in our storybook guide.

Only limitation, which is not optimal is difficult use of tailwindcss from plugins, because if we would include all the classes in ojs, it would result in very large bundle. Therefore only classes that are actually used in ojs would be possible to use in plugins, which is very unpredictable.

Option for very large plugin would be to have own tailwindcss build with own prefix to avoid any conflicts. But that also seems unnecessarily complex to setup in most cases.

Given that plugins ideally should be leveraging components from ui-component. The need for additional styling should be quite small. And recommended approach will be just to use css variables to pick up any font/color/sizing patterns from tailwindcss.  In Tailwindcss v4 css variables are going to be native part of the system and these can be easily used in plugins.  If v4 does not have stable release before releasing 3.5 - we can easily include these css variables to the config (already tested and works fine).

For reader side and themes, tailwindcss is unlikely be good fit as we aim for maximum flexibility and customisability. Therefore collocating tailwindcss classes with markup is not going to be good fit.  But there still would be benefits to allow using tailwindcss through @apply or some other mechanism without collocating. Thats to be determined once we get to theme technical refresh later on.
